In [1]:
import numpy as np
from numpy import linalg as LA


### Задача 1
Вкладчик положил 50000 условных единиц на три разных счета в три разных банка.

По первому счету выплаты составят 5% годовых, по второму – 7% годовых и по третьему 6% годовых. 

Известно, что через год вкладчик получил по процентам сумму в 2250 у.е. с первого и второго банка

И сумму в 1400 у.е. с первого и третьего банка.

Сколько условных единиц он положил на каждый счет изначально?

---------------------
Изначальные вклады в

Первый счет = x   |5% = 0.05

Второй счет = y   |7% = 0.07

Третий счет = z   |6% = 0.06

-------------------
x + y + z = 50000

0.05*x+0.07*y+0*z = 2250

0.05*x+0*y+0.06*z = 1400

x,y,z =?


In [2]:
a =[[1,1,1],
   [0.05,0.07,0],
   [0.05,0,0.06]]
b = [50000,2250,1400]
x = np.matmul(LA.inv(a),b)
x

array([10000., 25000., 15000.])

### Ответ задачи 1
Первый счет = 10000

Второй счет = 25000

Третий счет = 15000

### Задача 2 
На складе хранятся мобильные телефоны iPhone 6, iPhone 11, iPhone 12 моделей.

Всего 1328 экземпляра.

Моделей iPhone 6 на 120 меньше, чем моделей iPhone 11, и на 100 больше, чем моделей iPhone 12.

Сколько моделей каждого вида на складе?

---------------------
iPhone 6 = x

iPhone 11 = y 

iPhone 12 = z


x + y + z = 1328 

x - y - 0z = -120  

x - 0y - z = 100


In [3]:
# решение с помощью метода linalg.solve:
a = [[1,1,1],
     [1,-1,0],
     [1,0,-1]]
b =[1328,-120,100]
x = LA. solve(a,b)
x

array([436., 556., 336.])

In [4]:
# Проверка: 
np.dot(a, x)

array([1328., -120.,  100.])

In [5]:
# Решение через обратную матрицу:
# Множим обратную матрицу коєфициентов на столбец свободных членов
a = [[1,1,1],
     [1,-1,0],
     [1,0,-1]]
b =[1328,-120,100]
np.matmul(LA.inv(a),b)

array([436., 556., 336.])

### Ответ задачи 2:
iPhone 6 = 436

iPhone 11 = 556

iPhone 12 = 336

### Задача 3
Составьте каноническое уравнение эллипсоида, проходящего через три точки A,B,CA,B,C



### Задача 3:
Составьте каноническое уравнение эллипсоида, проходящего через три точки A,B,C

A=(sqrt(3), 0, sqrt(3))

B=(sqrt(6), 0.5, 0)

C=(1, 1/sqrt(3), 1)

Формула канонического уравнения элипсоида:

(x^2)/(a^2) + (y^2)/(b^2) + (z^2)/(c^2) = 1


Подставив в формулу значения точек, получим систему уравнений:

3/(a^2) + 0/(b^2) + 3/(c^2) = 1

6/(a^2) + 0.5/(b^2) + 0/(c^2) = 1

1/(a^2) + 0.33(3)/b^2 + 1/(c^2) = 1


Для того, что бы решить эту систему уравнений матричным способом, нужно привести каждое уравнение к виду
ax + by + cz = n

Нужно избавиться от знаменателей
Для этого, выразим их как коефициенты

k1 = 1/(a^2)

k2 = 1/(b^2)

k3 = 1/(c^2)

Теперь выражения будут иметь вид xk1 + yk2 + zk3 = 1

Стоит отметить, что в данном случае неизвестны k1,k2,k3
x,y,z - известны

Поэтому в данном случае x, y, z - это коэфициенты, а k1,k2,k3 - неизвестные
a - матрица коефициентов неизвестных 
b - матрица свободных членов

In [6]:
a = [[3,0,3],
     [6,0.5,0],
     [1,1/3,1]]
b = [1,1,1]
# x = np.matmul(LA.inv(a),b)
k = LA.solve(a,b)
k

array([0.        , 2.        , 0.33333333])

In [7]:
k1 = 0
k2 = 2
k3 = 1/3

k1: 

0 = 1/(a^2)

a= 0

-----------------------
k2: 

2 = 1/(b^2) | *b^2

2b^2 = 1 | /2

b^2 = 0,5

b = sqrt(0,5)

-----------------------
k3: 

1/3 = 1/(c^2) |*3c^2

c^2 = 3

c = sqrt(3)



### Ответ задачи 3

a= 0

b = sqrt(0,5)

c = sqrt(3)

# Задача 4 

Составьте уравнение параболы проходящей через заданные три точки (1,12), (3,54), (-1,2).

Необходимо найти коэффициенты уравнения a,b,c

y=ax^2 + bx + c

Подставив значения точек, получим систему уравнений:

1a + 1b + 1c = 12

9a + 3b + 1c = 54 

1a - 1b + 1c = 2 

Матрица коефициентов - а
Матрица свободных членов - b

In [8]:
a = [[1,1,1],
    [9,3,1],
    [1,-1,1]]

b =[12,54,2]
x = LA.solve(a,b)
x

array([4., 5., 3.])

### Ответ задачи 4

a = 4

b = 5

c = 3


### Задача 5
Составьте функцию get_polynom, которая по набору точек вида (x1,y1),(x2,y2),...., (xn+1,yn+1)

И будет находить коэффициенты многочлена Ci

Таким образом функция принимает список кортежей с координатами и возвращает набор коэффициентов

C0, C1, Cn

-------------
Итак, нужно написать функцию, которая может решить любое уравнение вида

Сi*X^i + (Ci-1) * X^(i-1)...  C1*X + C0) = Y

Например:

    3х^3 + 2x^2+ x = Y
    
    2x^2+ x = Y
    
    5x^5 + 4x^4 + 3x^3 + 2x^2 + x = Y
И т.д.

Соотвветственно, у нас должны быть матрицы А и В

А - Матрица коефициентов ( в данном случае координаты - Х)

В - Матрица свободных членов ( в данном случае - координаты У)

а исщутся в данном случае коефициенты (которые в примере выше были известными числами)

------
в матрице А нужно реализовать поднесение к степени,

так что бы первый елемент имел степень равную количестве пар координат в аргументе

Второй - степень на единицу меньшую, ... и последний - степень 1

In [20]:
coords =[[1,3],[5,2],[7,21],[23,1]] 
def get_polynom(coords):
    b = [] # матрица свободных членов (У)
    a = [] # матрица коефициентов (Х)
    for coord in coords:
        b.append(coord[1]) 
        
        for c in range(len(coords)):
            a.append(coord[0]**c) # Форирование матрицы с иксами, поднесенными к нужной степени 
    a=np.array(a)
    a=np.reshape(a, (len(coords), len(coords))) # Делаем из одномерной матрицы квадратную
    b = np.array(b)
    return np.linalg.solve(a, b)
get_polynom(coords)

array([ 14.91035354, -14.74747475,   2.93813131,  -0.1010101 ])